packages/languages/src/node/language-server-contribution.ts

```diff
@@ -7,33 +7,55 @@
 
 import * as net from 'net';
 import * as cp from 'child_process';
-import { injectable } from "inversify";
+import { injectable, inject } from "inversify";
 import { Message, isRequestMessage } from 'vscode-ws-jsonrpc';
 import { InitializeParams, InitializeRequest } from 'vscode-languageserver-protocol';
 import {
     createProcessSocketConnection,
-    createProcessStreamConnection,
+    createStreamConnection,
     forward,
     IConnection
 } from 'vscode-ws-jsonrpc/lib/server';
 import { MaybePromise } from "@theia/core/lib/common";
 import { LanguageContribution } from "../common";
+import { RawProcess, RawProcessFactory } from '@theia/process/lib/node/raw-process';
+import { ProcessManager } from '@theia/process/lib/node/process-manager';
 
 export {
     LanguageContribution, IConnection, Message
-}
+};
 
 export const LanguageServerContribution = Symbol('LanguageServerContribution');
 export interface LanguageServerContribution extends LanguageContribution {
     start(clientConnection: IConnection): void;
+    /**
+     * Called on shut down. Only synchronous operations are permitted.
+     */
+    stop?(): void;
 }
 
 @injectable()
 export abstract class BaseLanguageServerContribution implements LanguageServerContribution {
+
     abstract readonly id: string;
     abstract readonly name: string;
+    protected readonly spawnedProcessIds: number[];
```

Comment:
I wonder processes can exit even if the worker is still running, then they should be removed from this list. Maybe we can use disposable pattern instead of onStop callback, here disposable collection will be used which is returned as a disposable by start method.

Reply:
While this will change how the processes are disposed they will still be in the disposable list , even if they exited no ?

However this doesn't matter much.. if it has exited already the processManager will just ignore it and since ids are unique and not reused this is safe.

Reply:
I thought that will be removed on exit by calling dispose on disposable returned from collection.push.

Reply:
If you mean to make the Process disposable and that it disposes of itself on exit() that won't work since you can't be certain that all the event listeners on the object have been called once you dispose of it in your exit() handler.

Also it complicates things a lot because other components could have a reference to that process and try to call invalid methods on it now that it is disposed....

Or maybe I'm missing something?

Reply:
Ok, I only meant to consider an alternative api for onStop: use disposable as a return type for onStart. If it makes things complicated I am fine with the current state like now.